### DISCUSSION TITLE
Data science tools

### DISCUSSION THREAD

**Discussion by derekmahar** [dathere/qsv#785] [General]
(Continuation of [discussion](https://github.com/jqnatividad/qsv/issues/779#issuecomment-1434999453) about data science tools.)

How common and useful is [Apache Arrow](https://arrow.apache.org/) in data science?

**Comment by kindly** [comment]
Apache Arrow is more an underlying technology for many higher level tools like https://github.com/pola-rs/polars and https://arrow.apache.org/datafusion/ 

In my opinion, the tech is still a bit young, but is great if you are doing interoperability between languages.  I think even duckdb-rust uses arrow for passing data between the c++ library and rust.

**Comment by jqnatividad** [comment]
I've been eyeing [pola.rs](https://pola.rs) for a[ while](https://github.com/jqnatividad/qsv/issues/480), and looking to leverage it as an engine for qsv.

@baggiponte also suggested using it - https://github.com/jqnatividad/qsv/discussions/725

Right now, I'm thinking of tipping my toes into it by trying to use it for its [time-series capabilities](https://pola-rs.github.io/polars-book/user-guide/howcani/timeseries/intro.html), particularly it temporal groupby and resampling capabilities, as we often need to do so with CSV files from IoT type feeds. 

And it'd be very convenient to do so from the command-line, without having to write a script/program.

Are there any other "data-science" type CSV data-wrangling tasks from the command-line on your wishlist?

**Reply by derekmahar** [reply]
I'm highly interested in an asof join command similar to Polars' [asof join](https://pola-rs.github.io/polars-book/user-guide/howcani/combining_data/joining.html#asof-join) operation and Pandas' [`merge_asof`](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.merge_asof.html) function. The DuckDB team has [discussed](https://github.com/duckdb/duckdb/discussions/3207) implementing an asof join in DuckDB, but has no concrete schedule.  Fortunately, there is a workaround using a ["range intersection join"](https://github.com/duckdb/duckdb/discussions/3207#discussioncomment-2333697).  In 2018, in the Miller project, I unwittingly [requested](https://github.com/johnkerl/miller/issues/173) an asof join.  In that feature request, I mentioned the same SQL "range intersection join" solution that some members of the DuckDB team again [discussed](https://github.com/duckdb/duckdb/discussions/3207#discussioncomment-4669242).  Since I made that request, John Kerl, the Miller creator and maintainer, was busy re-implementing Miller in Go, so he hasn't yet had time to implement my feature request.  Should be any day now! ;)

In the meantime, I think qsv would benefit immensely from an asof join command! :)

**Reply by derekmahar** [reply]
Had I known earlier about Python Panda's [`merge_asof`](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.merge_asof.html), I probably would have started using Python much sooner!

**Reply by derekmahar** [reply]
It's funny.  In 2018, I started a quest to find a tool that could perform an asof join between two CSV files simply because I wanted to calculate the value of USD stock and cryptocurrency trades in CAD.  Most accountants would perform this calculation using Excel and the [VLOOKUP function](https://support.microsoft.com/en-us/office/vlookup-function-0bbc8083-26fe-4963-8ab8-93a18ad188a1), but I'm not an accountant.  Since then, I've experimented with [Miller](https://miller.readthedocs.io/en/latest/), [csvq](https://mithrandie.github.io/csvq/), [SQLite](https://sqlite.org/), [data.table](https://www.rdocumentation.org/packages/data.table/versions/1.14.6) in R, [DuckDB](https://duckdb.org/), and [Pandas](https://pandas.pydata.org/) in Python.  So far, DuckDB looks like the most promising tool to solve my asof join problem for larger data sets.

**Reply by derekmahar** [reply]
If my understanding of the concept is correct, I think I might use a temporal groupby to calculate the average price of a security during the 30 minute period before the expiration of a particular trading contract which happens to fall on hourly boundaries.  I've used DuckDB to perform this calculation simply by running an SQL average groupby query on timestamps truncated to hour precision and including all prices in the average that occurred within the 30 minute window before the hour.
